### PR TITLE
[storage fee] adjustments

### PR DESCRIPTION
**Pull request by msmouse**
### Description

[storage fee] take into account free txn and event bytes quota
[gas] 100x the GAS_SCALING_FACTOR, lowering execution gas charge by 100x

### Test Plan

existing test suite

### CODE REVIEW

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 1]
```diff
@@ -113,6 +113,11 @@ crate::params::define_gas_parameters!(
             { 5.. => "free_write_bytes_quota" },
             1024, // 1KB free per state write
         ],
+        [
+            free_event_bytes_quota: NumBytes,
+            { 7.. => "free_event_bytes_quota" },
+            1024, // 1KB free event bytes per transaction
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction/mod.rs:119]: Are you planning to benchmark 2KB after v1.4 cut? This is still fine actually as it'll be another week or two before we upgrade the gas schedule in testnet

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 2]
```diff
@@ -222,14 +227,21 @@ impl TransactionGasParameters {
         &self,
         events: impl IntoIterator<Item = &'a ContractEvent>,
     ) -> Fee {
-        events.into_iter().fold(Fee::zero(), |acc, event| {
-            acc + self.storage_fee_per_event_byte * NumBytes::new(event.size() as u64)
-        })
+        let total_bytes = events.into_iter().fold(NumBytes::zero(), |total, event| {
+            total + NumBytes::new(event.size() as u64)
+        });
+        total_bytes
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction/mod.rs:233]: Shouldn't this change be gated for replayability?

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 3]
```diff
@@ -222,14 +227,21 @@ impl TransactionGasParameters {
...
+            .checked_sub(self.free_event_bytes_quota)
+            .unwrap_or(NumBytes::zero())
+            * self.storage_fee_per_event_byte
     }
 
     /// New formula to charge storage fee for transaction based on fixed APT costs.
     pub fn calculate_transaction_storage_fee(&self, txn_size: NumBytes) -> Fee {
-        self.storage_fee_per_transaction_byte * txn_size
+        txn_size
+            .checked_sub(self.large_transaction_cutoff)
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction/mod.rs:242]: Why are we enforcing large_transaction_cutoff now?

File: aptos-move/e2e-move-tests/src/tests/per_category_gas_limits.rs [hunk 4]
```diff
@@ -46,10 +47,8 @@ fn io_limit_reached() {
     let acc = h.new_account_at(AccountAddress::from_hex_literal("0xbeef").unwrap());
     assert_success!(h.publish_package(&acc, &common::test_dir_path("execution_limit.data/test"),));
 
-    // Lower the max io gas to 10 units.
-    h.modify_gas_schedule(|gas_params| {
-        gas_params.txn.max_io_gas = Gas::new(10) * gas_params.txn.gas_unit_scaling_factor
-    });
+    // Lower the max io gas to lower than a single load_resource
+    h.modify_gas_schedule(|gas_params| gas_params.txn.max_io_gas = InternalGas::new(300_000 - 1));
```
Comment (movekevin) [inline aptos-move/e2e-move-tests/src/tests/per_category_gas_limits.rs:51]: What is 300_000 - 1?

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 1 above)
Comment (msmouse) [reply aptos-move/aptos-gas/src/transaction/mod.rs:119]: the 2KB is for the state write quota (`free_write_bytes_quota`), orthogonal to this.

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 2 above)
Comment (msmouse) [reply aptos-move/aptos-gas/src/transaction/mod.rs:233]: This is for the storage fee only, starting from gas feature version 7, never enabled before.

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 3 above)
Comment (msmouse) [reply aptos-move/aptos-gas/src/transaction/mod.rs:242]: To make smaller txns storage fee -free (virtually entire txn is close to free when the gas price is low). Again this is for the new "storage fee", although in this case `large_transaction_cutoff` has been respected for calculating the intrinsic gas https://github.com/aptos-labs/aptos-core/blob/bcdb4f2c98d448bce87151c97803ab7937f61ce0/aptos-move/aptos-gas/src/transaction/mod.rs#L251

File: aptos-move/e2e-move-tests/src/tests/per_category_gas_limits.rs (hunk 4 above)
Comment (msmouse) [reply aptos-move/e2e-move-tests/src/tests/per_category_gas_limits.rs:51]: 300_000 is the gas charge for a single load_resource (see the comment above)

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 2 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/transaction/mod.rs:233]: Is calculate_event_storage_fee already used? If so, wouldn't .checked_sub(self.free_event_bytes_quota) lead to new behavior?

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.exp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 processed 4 tasks
 
 task 1 'publish'. lines 4-30:
-Error: Transaction discarded. VM status code: MAX_GAS_UNITS_BELOW_MIN_TRANSACTION_GAS_UNITS
```
Comment (msmouse) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.exp:4]: These are so broken but I'd fix them separately.

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 2 above)
Comment (msmouse) [reply aptos-move/aptos-gas/src/transaction/mod.rs:233]: No all the "storage_fee" related functions and parameters were new with @vgao1996 's PR.

Comment (movekevin) [reply aptos-move/aptos-gas/src/transaction/mod.rs:233]: Got it. We'll run replay anyway I guess so we'll know if any issues happen due to missed gating.

**Review by movekevin — APPROVED** [review]

**Review by vgao1996 — APPROVED** [review]
LGTM

**Review by grao1991 — APPROVED** [review]

**Comment by hydrogenbond007** [comment]
hey great work, was curious is there a small md file or writeup on how this works?

**Comment by msmouse** [comment]
@hydrogenbond007  https://github.com/aptos-foundation/AIPs/issues/79

**Pull request merged**